Crash in WebKit::WebViewImpl::isInWindowFullscreenActive

eric-carlson opened the pull request:
#### 292a91632ec8b48d97e9fc5e82624661d5c0910c
<pre>
Crash in WebKit::WebViewImpl::isInWindowFullscreenActive
<a href="https://bugs.webkit.org/show_bug.cgi?id=275445">https://bugs.webkit.org/show_bug.cgi?id=275445</a>
<a href="https://rdar.apple.com/126704332">rdar://126704332</a>

Reviewed by Andy Estes.

Always null-check `WebPageProxy.playbackSessionManager()` because
there is no guarantee that it is valid.

* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::protectedPlaybackSessionInterface):
(WebKit::WebViewImpl::isInWindowFullscreenActive const):
(WebKit::WebViewImpl::toggleInWindowFullscreen):
(WebKit::WebViewImpl::updateMediaPlaybackControlsManager):

Canonical link: <a href="https://commits.webkit.org/280029@main">https://commits.webkit.org/280029@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ef6537c8947a46de35d41c30fcd162a1bd586d46

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55534 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34857 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8001 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58518 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5964 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42479 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6163 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44719 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4092 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57563 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32764 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47862 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25847 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29548 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4108 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/51483 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5458 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60109 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30688 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5581 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/loading-the-media-resource/resource-selection-invoke-remove-from-document-networkState.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52153 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31773 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47932 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51627 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12307 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32854 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31520 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -6188,17 +6188,26 @@ static NSTextAlignment nsTextAlignmentFromTextAlignment(TextAlignment textAlignm
+
+RefPtr<PlatformPlaybackSessionInterface> protectedPlaybackSessionInterface()
+{
+    if (auto* manager = m_page->playbackSessionManager())
```
```suggestion
    if (RefPtr manager = m_page->playbackSessionManager())
```

aestes commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -6188,17 +6188,26 @@ static NSTextAlignment nsTextAlignmentFromTextAlignment(TextAlignment textAlignm
 bool WebViewImpl::isInWindowFullscreenActive() const
 {
-    if (auto* interface = m_page->playbackSessionManager()->controlsManagerInterface())
+    if (auto* interface = protectedPlaybackSessionInterface())
```
```suggestion
    if (RefPtr interface = protectedPlaybackSessionInterface())
```

aestes commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -6188,17 +6188,26 @@ static NSTextAlignment nsTextAlignmentFromTextAlignment(TextAlignment textAlignm
 void WebViewImpl::toggleInWindowFullscreen()
 {
-    if (auto* interface = m_page->playbackSessionManager()->controlsManagerInterface())
+    if (auto* interface = protectedPlaybackSessionInterface())
```
```suggestion
    if (RefPtr interface = protectedPlaybackSessionInterface())
```

aestes commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -6213,7 +6222,7 @@ static NSTextAlignment nsTextAlignmentFromTextAlignment(TextAlignment textAlignm
     }
 
-    if (WebCore::PlatformPlaybackSessionInterface* interface = m_page->playbackSessionManager()->controlsManagerInterface()) {
+    if (auto* interface = protectedPlaybackSessionInterface()) {
```
```suggestion
    if (RefPtr interface = protectedPlaybackSessionInterface()) {
```

aestes reviewed (approved):
r=me with comments. we should make sure we're storing a RefPtr to the protected object on the stack.

eric-carlson committed:
Crash in WebKit::WebViewImpl::isInWindowFullscreenActive
https://bugs.webkit.org/show_bug.cgi?id=275445
rdar://126704332

Reviewed by Andy Estes.

Always null-check `WebPageProxy.playbackSessionManager()` because
there is no guarantee that it is valid.

* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::protectedPlaybackSessionInterface):
(WebKit::WebViewImpl::isInWindowFullscreenActive const):
(WebKit::WebViewImpl::toggleInWindowFullscreen):
(WebKit::WebViewImpl::updateMediaPlaybackControlsManager):

Canonical link: https://commits.webkit.org/280029@main
